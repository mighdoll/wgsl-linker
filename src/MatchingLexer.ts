import { tracing, parserLog } from "./ParserTracing.js";
import { Token, TokenMatcher } from "./TokenMatcher.js";

export interface Lexer {
  next(): Token | undefined;
  withMatcher<T>(tokenMatcher: TokenMatcher, fn: () => T): T;
  position(pos?: number): number;
  eof(): boolean;
}

export function matchingLexer(
  src: string,
  rootMatcher: TokenMatcher,
  ignore = new Set(["ws"])
): Lexer {
  let matcher = rootMatcher;
  const matcherStack: TokenMatcher[] = [];

  matcher.start(src);

  function next(): Token | undefined {
    let token = matcher.next();
    while (token && ignore.has(token.kind)) {
      token = matcher.next();
    }
    tracing && parserLog(`: ${JSON.stringify(token?.text)} (${token?.kind}) ${matcher.position()}`);
    return token;
  }

  function pushMatcher(newMatcher: TokenMatcher): void {
    const position = matcher.position();
    matcherStack.push(matcher);
    newMatcher.start(src, position);
    matcher = newMatcher;
  }

  function popMatcher(): void {
    const position = matcher.position();
    matcher =
      matcherStack.pop() || (console.error("too many pops"), rootMatcher);
    matcher.position(position);
  }

  function position(pos?: number): number {
    if (pos !== undefined) {
      matcher.start(src, pos);
    }
    return matcher.position();
  }

  function withMatcher<T>(tokenMatcher: TokenMatcher, fn: () => T): T {
    pushMatcher(tokenMatcher);
    const result = fn();
    popMatcher();
    return result;
  }

  function eof(): boolean {
    return matcher.position() === src.length;
  }

  return {
    next,
    position,
    withMatcher,
    eof,
  };
}
