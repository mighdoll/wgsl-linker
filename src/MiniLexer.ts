import { tokenMatcher } from "./TokenMatcher.js";

/** token matchers for parrts of wgsl and #directives */
/*
syntax we aim to parse:

#import name <(arg1, arg2)> <from moduleName> <as rename>
#export <name> <(arg1, arg2)>
#replace 128=workgroupSize 
// 
/* * /

fn decl() 
struct decl { }

fnCall()
StructConstruct()
: StructType
<Struct, ...>  
*/

const exportD = "#export";
const importD = "#import";
const directives = { exportD, importD };
const word = /[a-zA-Z_][a-zA-Z0-9_]+/;
const ws = /\s+/;
const lineComment = "//";
const digits = /[0-9]+/;

/** matching tokens at wgsl root level */
export const mainMatch = tokenMatcher({
  ...directives,
  lineComment,
  commentStart: "/*",
  commentEnd: "*/",
  annotation: /@[a-zA-Z_][a-zA-Z0-9_]+/,
  word,
  ws,
});

const notDirective = /[^#]+$/;
const eol = /$/;

/** matching tokens at the start of a '//' line comment */
export const lineCommentMatch = tokenMatcher({
  ...directives,
  notDirective,
  eol,
});
const lparen = "(";
const rparen = ")";
const comma = ",";
const equals = "=";

/** matching tokens while parsing directive parameters #export foo(param1, param2) */
export const directiveArgsMatch = tokenMatcher({
  lparen,
  rparen,
  from: "from",
  as: "as",
  word,
  digits,
  ws,
  comma,
  equals,
  eol,
});

// const symbolSet =
//   "& && -> @ / ! [ ] { } : , = == != > >= >> < << <= % - --" +
//   ". + ++ | || () ; * ~ ^ += -= *= /= %= &= |= ^= >>= <<= <<";

// const symbolList = symbolSet.split(" ").sort((a, b) => b.length - a.length);
// const escaped = symbolList.map(escapeRegex);
// export const symbol = new RegExp(escaped.join("|"));

// export function regexOr(flags: string, ...exp: RegExp[]): RegExp {
//   const concat = exp.map((e) => e.source).join("|");
//   return new RegExp(concat, flags);
// }
