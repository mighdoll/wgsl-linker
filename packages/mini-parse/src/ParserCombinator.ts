import {
  CombinatorArg,
  OrParser,
  ParserFromArg,
  ParserFromRepeatArg,
  TagsFromArg,
  ResultFromArg,
  SeqParser,
  SeqValues,
  AsRecordArray,
} from "./CombinatorTypes.js";
import { quotedText } from "./MatchingLexer.js";
import {
  ExtendedResult,
  TagRecord,
  NoTags,
  OptParserResult,
  Parser,
  ParserContext,
  parser,
  runExtended,
  simpleParser,
  tokenSkipSet,
  ParserResult,
} from "./Parser.js";
import { ctxLog } from "./ParserLogging.js";
import { mergeTags } from "./ParserUtil.js";
import { Token, TokenMatcher } from "./TokenMatcher.js";

/** Parsing Combinators
 *
 * The basic idea is that parsers are contructed heirarchically from other parsers.
 * Each parser is independently testable and reusable with combinators like or() and seq().
 *
 * Each parser is a function that recognizes tokens produced by a lexer
 * and returns a result.
 *  Returning null indicate failure. Tokens are not consumed on failure.
 *  Users can also use the .tag() method to tag results from a stage. Tagged results
 *    propagate up to containing parsers for convenience in selecting results.
 *
 * Built in parsers and combinators are available:
 *  kind() recognizes tokens of a particular type.
 *  or(), seq(), opt(), map() and repeat() combine other stages.
 *
 * Users construct their own parsers by combining other parser stages
 * and typically use map() to report results. Results can be stored
 * in the array app[], which is provided by the user and available for
 * all user constructed parsers.
 */

export class ParseError extends Error {
  constructor(msg?: string) {
    super(msg);
  }
}

/** Parse for a particular kind of token,
 * @return the matching text */
export function kind(kindStr: string): Parser<string> {
  return simpleParser(
    `kind '${kindStr}'`,
    (state: ParserContext): string | null => {
      const next = state.lexer.next();
      return next?.kind === kindStr ? next.text : null;
    }
  );
}

/** Parse for a token containing a text value
 * @return the kind of token that matched */
export function text(value: string): Parser<string, NoTags> {
  return simpleParser(
    `text ${quotedText(value)}'`,
    (state: ParserContext): string | null => {
      const next = state.lexer.next();
      return next?.text === value ? next.text : null;
    }
  );
}

/** Parse a sequence of parsers
 * @return an array of all parsed results, or null if any parser fails */
export function seq<P extends CombinatorArg[]>(...args: P): SeqParser<P> {
  const parsers = args.map(parserArg);

  const result = parser("seq", (ctx: ParserContext) => {
    const values = [];
    let tagged = {};
    for (const p of parsers) {
      const result = p._run(ctx);
      if (result === null) return null;

      tagged = mergeTags(tagged, result.tags);
      values.push(result.value);
    }
    return { value: values, tags: tagged };
  });

  return result as SeqParser<P>;
}

export function test(): void {
  const a = seq(text("a").tag("AA"));
  const p = seq(seq(a));

  p.map((r) => {
    const s: string[] = r.tags.AA;
    console.log(s);
  });
}

/** Try parsing with one or more parsers,
 *  @return the first successful parse */
export function or<P extends CombinatorArg[]>(...args: P): OrParser<P> {
  const parsers = args.map(parserArg);
  const result = parser("or", (state: ParserContext) => {
    for (const p of parsers) {
      const result = p._run(state);
      if (result !== null) {
        return result;
      }
    }
    return null;
  });

  return result as OrParser<P>;
}

const undefinedResult: ParserResult<undefined, NoTags> = {
  value: undefined,
  tags: {},
};

export type UndefinedParser = Parser<undefined, NoTags>;
/** Try a parser.
 *
 * If the parse succeeds, return the result.
 * If the parser fails, return false and don't advance the input. Returning false
 * indicates a successful parse, so combinators like seq() will succeed.
 */
export function opt<P extends CombinatorArg>(
  arg: P
): ParserFromArg<P> | UndefinedParser {
  const p = parserArg(arg);

  const result: ParserFromArg<P> | UndefinedParser = parser(
    "opt",
    (state: ParserContext) => {
      const result = p._run(state);
      // If parsing fails, we return instead a success
      // with 'undefined' as a value

      // cast the undefined result here and recover type with the ascription above
      type PR = ParserResult<ResultFromArg<P>, TagsFromArg<P>>;
      return result || (undefinedResult as PR);
    }
  );
  return result;
}

/** return true if the provided parser _doesn't_ match
 * does not consume any tokens */
export function not(arg: CombinatorArg): Parser<true> {
  const p = parserArg(arg);
  return parser("not", (state: ParserContext) => {
    const pos = state.lexer.position();
    const result = p._run(state);
    if (!result) {
      return { value: true, tags: {} };
    }
    state.lexer.position(pos);
    return null;
  });
}

/** yield next token, any token */
export function any(): Parser<Token> {
  return simpleParser("any", (state: ParserContext): Token | null => {
    const next = state.lexer.next();
    return next || null;
  });
}

/** yield next token if the provided parser doesn't match */
export function anyNot(arg: CombinatorArg): Parser<Token> {
  return seq(not(arg), any())
    .map((r) => r.value[1])
    .traceName("anyNot");
}

/** match everything until a terminator (and the terminator too) */
export function anyThrough<A extends CombinatorArg>(
  arg: A
): Parser<[...any, ResultFromArg<A>], TagsFromArg<A>> {
  const p = parserArg<A>(arg);
  const result = seq(repeat(anyNot(p)), p).traceName(
    `anyThrough ${p.debugName}`
  );
  return result as any; // TODO fix any type
}

/** match zero or more instances of a parser */
export function repeat<A extends CombinatorArg>(
  arg: A
): ParserFromRepeatArg<A> {
  return parser("repeat", repeatWhileFilter(arg));
}

type ResultFilterFn<T> = (
  result: ExtendedResult<T | string, any>
) => boolean | undefined;

export function repeatWhile<A extends CombinatorArg>(
  arg: A,
  filterFn: ResultFilterFn<ResultFromArg<A>>
): ParserFromRepeatArg<A> {
  return parser("repeatWhile", repeatWhileFilter(arg, filterFn));
}

type RepeatWhileResult<A extends CombinatorArg> = OptParserResult<
  SeqValues<A[]>,
  TagsFromArg<A>
>;

function repeatWhileFilter<A extends CombinatorArg>(
  arg: A,
  filterFn: ResultFilterFn<ResultFromArg<A>> = () => true
): (ctx: ParserContext) => RepeatWhileResult<A> {
  const p = parserArg(arg);
  return (ctx: ParserContext): RepeatWhileResult<A> => {
    const values: ResultFromArg<A>[] = [];
    let results = {};
    for (;;) {
      const result = runExtended<ResultFromArg<A>, TagsFromArg<A>>(ctx, p);

      // continue acccumulating until we get a null or the filter tells us to stop
      if (result !== null && filterFn(result)) {
        values.push(result.value);
        results = mergeTags(results, result.tags);
      } else {
        // always return succcess
        const r = { value: values, tags: results };
        return r as any; // TODO typing of better tagged results
      }
    }
  };
}

/** A delayed parser definition, for making recursive parser definitions. */
// TODO make this private
export function fn<T, N extends TagRecord>(
  fn: () => Parser<T, N>
): Parser<T, N> {
  return parser("fn", (state: ParserContext): OptParserResult<T, N> => {
    const stage = fn();
    return stage._run(state);
  });
}

/** yields true if parsing has reached the end of input */
export function eof(): Parser<true> {
  return simpleParser(
    "eof",
    (state: ParserContext) => state.lexer.eof() || null
  );
}

/** if parsing fails, log an error and abort parsing */
export function req<A extends CombinatorArg>(
  arg: A,
  msg?: string
): ParserFromArg<A> {
  const p = parserArg(arg);
  return parser("req", (ctx: ParserContext) => {
    const result = p._run(ctx);
    if (result === null) {
      ctxLog(ctx, msg ?? `expected ${p.debugName}`);
      throw new ParseError();
    }
    return result as any; // TODO rm cast?
  });
}

/** always succeeds, does not consume any tokens */
export function yes(): Parser<true> {
  return simpleParser("yes", () => true);
}

/** always fails, does not consume any tokens */
export function no(): Parser<null> {
  return simpleParser("no", () => null);
}

export interface WithSepOptions {
  /** if true, allow an optional trailing separator (default true) */
  trailing?: boolean;
  /** if true, require at least one element (default false) */
  requireOne?: boolean;
}

/** match an optional series of elements separated by a delimiter (e.g. a comma) */
export function withSep<T, N extends TagRecord>(
  sep: CombinatorArg,
  p: Parser<T, N>,
  opts: WithSepOptions = {}
): Parser<T[], N> {
  const elem = Symbol();
  const { trailing = true, requireOne = false } = opts;
  const first = requireOne ? p : opt(p);
  const last = trailing ? opt(sep) : yes();

  return seq(first.tag(elem), repeat(seq(sep, p.tag(elem))), last)
    .map((r) => {
      const tags = r.tags as Record<symbol, T>;
      return tags[elem];
    })
    .traceName("withSep") as any;
}

/** run a parser with a provided token matcher (i.e. use a temporary lexing mode) */
export function tokens<A extends CombinatorArg>(
  matcher: TokenMatcher,
  arg: A
): ParserFromArg<A> {
  const p = parserArg(arg);
  return parser(`tokens ${matcher._traceName}`, (state: ParserContext) => {
    return state.lexer.withMatcher(matcher, () => {
      return p._run(state);
    });
  });
}

/** return a parser that matches end of line, or end of file,
 * optionally preceded by white space
 * @param ws should not match \n */
export function makeEolf(matcher: TokenMatcher, ws: string): Parser<any> {
  // prettier-ignore
  return tokens(matcher, 
      tokenSkipSet(null, // disable automatic ws skipping so we can match newline
        seq(
          opt(kind(ws)), 
          or("\n", eof())
        )
      )
    )
   .traceName("eolf");
}

/** convert naked string arguments into text() parsers and functions into fn() parsers */
export function parserArg<A extends CombinatorArg>(arg: A): ParserFromArg<A> {
  if (typeof arg === "string") {
    return text(arg) as ParserFromArg<A>; // LATER fix cast
  } else if (arg instanceof Parser) {
    return arg as Parser<ResultFromArg<A>, TagsFromArg<A>>;
  }
  return fn(arg as () => ParserFromArg<A>);
}
