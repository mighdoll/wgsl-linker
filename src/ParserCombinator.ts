import { Lexer } from "./MiniLexer.js";

export interface ParserContext {
  lexer: Lexer;
  results: any[];
}

export type ParserStage<T> = (state: ParserContext) => T;
export type ParserFn<T> = (state: ParserContext) => T;

export function parserStage<T>(fn: ParserFn<T>): ParserStage<T> {
  // add fluent interface methods?
  // reset position on failure?
  // return start position
  return fn;
}

export const kind = (kind: string) =>
  parserStage((state: ParserContext): string | null => {
    const next = state.lexer.next();
    return next?.kind === kind ? next.text : null;
  });

export function or<T, U>(
  a: ParserStage<T>,
  b: ParserStage<U>
): ParserStage<T | U> {
  return parserStage((state: ParserContext) => {
    const pos = state.lexer.position();
    const aResult = a(state);
    if (aResult) {
      return aResult;
    }
    state.lexer.setPosition(pos);
    return b(state);
  });
}

export function seq<T, U>(
  a: ParserStage<T | null>,
  b: ParserStage<U | null>
): ParserStage<[T, U] | null>;
export function seq<T, U, V>(
  a: ParserStage<T>,
  b: ParserStage<U>,
  c: ParserStage<V>
): ParserStage<[T, U, V] | null>;
export function seq(...stages: ParserStage<any>[]): ParserStage<any[] | null> {
  return parserStage((state: ParserContext) => {
    const results = [];
    for (const stage of stages) {
      const result = stage(state);
      if (result === null || result === undefined) {
        return null;
      }
      results.push(result);
    }
    return results;
  });
}

/*
  
   // syntax 1. wrapper functions for or, seq, opt, rep
  const root = or(lineComment, fnDecl);
  const lineComment = seq(kind("lineComment", opt(directive)));
  const results = parse(root, lexer)

  // syntax 2. fluent interface 
  const root = lineComment.or(fnDecl);
  const lineComment = kind("lineComment").opt(directive);

  in either case, each stage needs to take in a ParserContext.
  Output could be a struct, or maybe just a non-undefined value.

  TS type for the result value for some of these combinators is tricky, 
  something like a seq returns an array of multiple types..
  
*/

function parserChain(src: string) {}
