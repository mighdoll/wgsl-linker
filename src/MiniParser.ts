import { Lexer, matchingLexer } from "./MatchingLexer.js";
import {
  directiveArgsMatch,
  lineCommentMatch,
  mainMatch,
} from "./MiniLexer.js";
import { ParserContext, kind, or, parsing, seq } from "./ParserCombinator.js";

interface ParserState {
  lexer: Lexer;
  results: AbstractElem[];
}

export type AbstractElem = DirectiveElem;

/** 'interesting' elements found in the source */
export interface AbstractElemBase {
  kind: string;
  position: number;
}

export interface DirectiveElem extends AbstractElemBase {
  kind: "directive";
  name: string;
  args: string[];
}

const singleWord = parsing((state: ParserState): string | null => {
  return state.lexer.withMatcher(directiveArgsMatch, () => {
    const x = seq(kind("word"), kind("eol"))(state);
    return x?.value[0] || null;
  });
});

const directive = parsing((state: ParserState): string | null => {
  const directiveElems = seq(kind("directive"), singleWord)(state);
  if (directiveElems) {
    const [direct, word] = directiveElems.value;
    const name = direct;
    const position = state.lexer.position();
    state.results.push({ kind: "directive", name, args: [word], position });
    return name;
  } else {
    return null;
  }
});

export const lineComment = parsing((state: ParserState): boolean | null => {
  return state.lexer.withMatcher(lineCommentMatch, () => {
    const afterComment = or(directive, kind("notDirective"));
    const parser = seq(kind("lineComment"), afterComment);
    return parser(state) === null ? null : true;
  });
});

const root = or(directive, lineComment);

export function miniParse(src: string): AbstractElem[] {
  const lexer = matchingLexer(src, mainMatch);
  const results: AbstractElem[] = [];

  const state: ParserContext = { lexer, results };
  root(state);

  return state.results;
}

// function parenArgs(): string[] | null {
//   return lexer.withMatcher(directiveArgsMatch, () => {
//     const results = seq("lparen", "word", "rparen", "eol");
//     if (!results) {
//       return null;
//     }
//     const args = [results[1]];
//     return args;
//   });
// }

// export const lineComment = parsing(
//   (state: ParserState): OptParserResult<any> => {
//     return state.lexer.withMatcher(lineCommentMatch, () => {
//       const afterComment = or(directive, kind("notDirective"));
//       const parser = seq(kind("lineComment"), afterComment);
//       return parser(state);
//     });
//   }
// );
