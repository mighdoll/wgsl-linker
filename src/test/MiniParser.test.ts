import { expect, test } from "vitest";
import { matchingLexer } from "../MatchingLexer.js";
import { mainMatch } from "../MiniLexer.js";
import { directive, lineComment, miniParse } from "../MiniParser.js";
import { ParserContext, ParserStage } from "../ParserCombinator.js";

test("parse empty string", () => {
  const parsed = miniParse("");
  expect(parsed).toMatchSnapshot();
});

test("directive parses #export", () => {
  const parsed = testParse(directive, "#export");
  expect(parsed.app[0].kind).equals("export");
});

test("parse #export", () => {
  const parsed = miniParse("#export");
  expect(parsed[0].kind).equals("export");
});

test("parse #export foo", () => {
  const parsed = miniParse("#export foo");
  expect(parsed).toMatchSnapshot();
});

test("parse #export foo(bar)", () => {
  const parsed = miniParse("#export foo(bar)");
  expect(parsed).toMatchSnapshot();
});

test("parse #export foo(bar, baz, boo)", () => {
  const parsed = miniParse("#export foo(bar, baz, boo)");
  expect(parsed).toMatchSnapshot();
});

test("parse #import foo", () => {
  const parsed = miniParse("#import foo");
  expect(parsed).toMatchSnapshot();
});

test("parse #import foo(a,b) from bar as baz", () => {
  const parsed = miniParse("#import foo from bar as baz");
  expect(parsed).toMatchSnapshot();
});

test("lineComment parse // foo bar", () => {
  const src = "// foo bar";
  const state = testParse(lineComment, src);
  expect(state.lexer.position()).eq(src.length);
});

test("lineComment parse // #export foo", () => {
  const src = "// #export foo";
  const state = testParse(lineComment, src);
  expect(state.lexer.position()).eq(src.length);
  expect(state.app).toMatchSnapshot();
});

test("parse fn foo() {}", () => {
  const src = "fn foo() { }";
  const parsed = miniParse(src);
  expect(parsed).toMatchSnapshot();
});

export function testParse<T>(
  stage: ParserStage<T>,
  src: string
): ParserContext {
  const lexer = matchingLexer(src, mainMatch);
  const app: any[] = [];

  const state: ParserContext = { lexer, app };
  stage(state);

  return state;
}
