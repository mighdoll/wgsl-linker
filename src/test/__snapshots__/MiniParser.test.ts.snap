// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lineComment parse // #export foo 1`] = `
[
  {
    "args": undefined,
    "end": 14,
    "kind": "export",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse #export foo 1`] = `
[
  {
    "args": undefined,
    "end": 11,
    "kind": "export",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse #export foo(bar) 1`] = `
[
  {
    "args": [
      [
        "bar",
      ],
    ],
    "end": 16,
    "kind": "export",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse #export foo(bar, baz, boo) 1`] = `
[
  {
    "args": [
      [
        "bar",
        "baz",
        "boo",
      ],
    ],
    "end": 26,
    "kind": "export",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse #import foo 1`] = `
[
  {
    "args": undefined,
    "as": undefined,
    "end": 11,
    "from": undefined,
    "kind": "import",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse #import foo(a,b) from bar as baz 1`] = `
[
  {
    "args": undefined,
    "as": "baz",
    "end": 27,
    "from": "bar",
    "kind": "import",
    "name": "foo",
    "start": 0,
  },
]
`;

exports[`parse empty string 1`] = `[]`;
